subsys/nrf_security/src/drivers/cracen/cracenpsa/src/key_management.c

```diff
@@ -38,79 +38,101 @@ extern const uint8_t cracen_N3072[384];
 
 extern nrf_security_mutex_t cracen_mutex_symmetric;
 
-static psa_status_t check_brainpool_alg_and_key_bits(psa_algorithm_t alg, size_t key_bits)
+#define CHECK_BITS(key_bits, arr)                                                                  \
+	for (size_t check_bits_i = 0; check_bits_i < ARRAY_SIZE(arr); check_bits_i++) {            \
+		if ((arr)[check_bits_i] == *(key_bits))                                            \
+			return PSA_SUCCESS;                                                        \
+	}
+
+/* Assumes arr is sorted. */
+#define GET_BITS(key_buffer_size, arr)                                                             \
+	for (size_t get_bits_i = ARRAY_SIZE(arr); get_bits_i > 0; get_bits_i--) {                  \
+		if (PSA_BITS_TO_BYTES((arr)[get_bits_i - 1]) <= (key_buffer_size)) {               \
+			*key_bits = (arr)[get_bits_i - 1];                                         \
+			return PSA_SUCCESS;                                                        \
+		}                                                                                  \
+	}
+
+static psa_status_t check_brainpool_alg_and_key_bits(psa_algorithm_t alg, size_t key_buffer_size,
+						     size_t *key_bits)
 {
 	if (!PSA_ALG_IS_ECDSA(alg) && alg != PSA_ALG_ECDH) {
 		return PSA_ERROR_INVALID_ARGUMENT;
 	}
 
-	switch (key_bits) {
-	case 192:
-	case 224:
-	case 256:
-	case 320:
-	case 384:
-	case 512:
-		return PSA_SUCCESS;
-	default:
-		return PSA_ERROR_INVALID_ARGUMENT;
+	size_t valid_bits[] = {192, 224, 256, 320, 384, 512};
+
+	if (*key_bits == 0) {
+		GET_BITS(key_buffer_size, valid_bits);
 	}
+	CHECK_BITS(key_bits, valid_bits);
+
+	return PSA_ERROR_INVALID_ARGUMENT;
 }
 
-static psa_status_t check_secp_k1_alg_and_key_bits(psa_algorithm_t alg, size_t key_bits)
+static psa_status_t check_secp_k1_alg_and_key_bits(psa_algorithm_t alg, size_t key_buffer_size,
+						   size_t *key_bits)
 {
 	if (!PSA_ALG_IS_ECDSA(alg) && alg != PSA_ALG_ECDH) {
 		return PSA_ERROR_INVALID_ARGUMENT;
 	}
 
-	switch (key_bits) {
-	case 192:
-	case 256:
-		return PSA_SUCCESS;
-	case 225:
-		return PSA_ERROR_NOT_SUPPORTED;
-	default:
-		return PSA_ERROR_INVALID_ARGUMENT;
+	size_t valid_bits[] = {192, 225, 256};
+
+	if (*key_bits == 0) {
+		GET_BITS(key_buffer_size, valid_bits);
 	}
+	CHECK_BITS(key_bits, valid_bits);
+
+	return PSA_ERROR_INVALID_ARGUMENT;
```

Comment:
Here we return invalid argument even for the 225 which used to be not-supported right? It's probably an edge case but did you intend to do this?

Reply:
tried simplifying things here a bit